enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,114 @@
+(ns metabase-enterprise.caching.strategies
+  (:require
+   [java-time.api :as t]
+   [malli.core :as mc]
+   [metabase.public-settings :as public-settings]
+   [metabase.public-settings.premium-features :refer [defenterprise defenterprise-schema]]
+   [metabase.query-processor.middleware.cache-backend.db :as backend.db]
+   [metabase.util.cron :as u.cron]
+   [metabase.util.log :as log]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(def CacheStrategy ;; not used in API since there are some additional fields
+  "Schema for a caching strategy"
+  (mc/schema
+   [:and
+    [:map
+     [:type [:enum :nocache :ttl :duration :schedule :query]]]
+    [:multi {:dispatch :type}
+     [:nocache  [:map
+                 [:type keyword?]]]
+     [:ttl      [:map
+                 [:type keyword?]
+                 [:multiplier ms/PositiveInt]
+                 [:min_duration ms/PositiveInt]]]
+     [:duration [:map
+                 [:type keyword?]
+                 [:duration ms/PositiveInt]
+                 [:unit [:enum "hours" "minutes" "seconds" "days"]]]]
+     [:schedule [:map
+                 [:type keyword?]
+                 [:last-expired-at [:maybe some?]]
+                 [:schedule u.cron/CronScheduleString]]]
+     [:query    [:map
+                 [:type keyword?]
+                 [:last-expired-at [:maybe some?]]
+                 [:field_id int?]
+                 [:aggregation [:enum "max" "count"]]
+                 [:schedule u.cron/CronScheduleString]]]]]))
+
+(defn granular-duration-hours
+  "Returns the granular cache duration (in hours) for a card. On EE, this first checking whether there is a stored value
+   for the card, dashboard, or database (in that order of decreasing preference). Returns nil on OSS."
+  [card dashboard-id]
+  (or (:cache_ttl card)
+      (:cache_ttl (t2/select-one [:model/Dashboard :cache_ttl] :id dashboard-id))
+      (:cache_ttl (t2/select-one [:model/Database :cache_ttl] :id (:database_id card)))))
+
+(defenterprise-schema granular-cache-strategy :- [:maybe CacheStrategy]
+  "Returns the granular cache strategy for a card."
+  :feature :cache-granular-controls
+  [card dashboard-id]
+  (when (public-settings/enable-query-caching)
+    (let [qs     (for [[i model model-id] [[1 "question"   (:id card)]
+                                           [2 "dashboard"  dashboard-id]
+                                           [3 "database"   (:database_id card)]
+                                           [4 "root"       0]]
+                       :when              model-id]
+                   {:from   [:cache_config]
```

Comment:
nit: could you reorder the key of these hsql forms so it's consistent? my preference is select -> from -> where -> limit

Reply:
This is intentional, to know what `select` contains you have to know where it's coming from.